File: FakturusTrack.Android/app/src/main/java/com/fakturus/track/features/timetracking/TimeTrackingViewModel.kt
package com.fakturus.track.features.timetracking

import android.content.SharedPreferences
import android.util.Log
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.fakturus.track.models.AppDatabase
import com.fakturus.track.models.PendingDeleteEntity
import com.fakturus.track.models.WorkSessionEntity
import com.fakturus.track.services.api.APIClient
import com.fakturus.track.services.sync.SyncEngine
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.time.Duration
import java.time.Instant
import java.time.LocalDate
import kotlin.math.ceil

class TimeTrackingViewModel(
    private val database: AppDatabase,
    private val syncEngine: SyncEngine? = null,
    private val apiClient: APIClient? = null,
    private val prefs: SharedPreferences? = null
) : ViewModel() {

    private val dao = database.workSessionDao()

    private val _activeSession = MutableStateFlow<WorkSessionEntity?>(null)
    val activeSession: StateFlow<WorkSessionEntity?> = _activeSession.asStateFlow()

    val sessions: Flow<List<WorkSessionEntity>> = dao.getAllOrderedByDate()

    private val _isLoading = MutableStateFlow(false)
    val isLoading: StateFlow<Boolean> = _isLoading.asStateFlow()

    private val _error = MutableStateFlow<String?>(null)
    val error: StateFlow<String?> = _error.asStateFlow()

    // Pause state (E08)
    private val _isPaused = MutableStateFlow(false)
    val isPaused: StateFlow<Boolean> = _isPaused.asStateFlow()

    private val _currentPauseStart = MutableStateFlow<Instant?>(null)
    val currentPauseStart: StateFlow<Instant?> = _currentPauseStart.asStateFlow()

    private val _accumulatedPauseMinutes = MutableStateFlow(0)
    val accumulatedPauseMinutes: StateFlow<Int> = _accumulatedPauseMinutes.asStateFlow()

    // ArbZG banner dismiss tracking
    private val _hasShown6h = MutableStateFlow(false)
    val hasShown6h: StateFlow<Boolean> = _hasShown6h.asStateFlow()
    private val _hasShown9h = MutableStateFlow(false)
    val hasShown9h: StateFlow<Boolean> = _hasShown9h.asStateFlow()
    private val _hasShown10h = MutableStateFlow(false)
    val hasShown10h: StateFlow<Boolean> = _hasShown10h.asStateFlow()

    // Expose sync state
    val isSyncing: StateFlow<Boolean> = syncEngine?.isSyncing ?: MutableStateFlow(false)
    val syncLastError: StateFlow<String?> = syncEngine?.lastError ?: MutableStateFlow(null)

    init {
        viewModelScope.launch {
            try {
                val session = dao.getActiveSession()
                _activeSession.value = session

                // Crash recovery: check if a pause was running
                if (session != null) {
                    _accumulatedPauseMinutes.value = session.pauseMinutes
                    val savedPauseStart = prefs?.getString("currentPauseStart", null)
                    if (savedPauseStart != null) {
                        try {
                            _currentPauseStart.value = Instant.parse(savedPauseStart)
                            _isPaused.value = true
                        } catch (_: Exception) {
                            prefs.edit().remove("currentPauseStart").apply()
                        }
                    }
                }
            } catch (e: Exception) {
                _error.value = "Fehler beim Laden der aktiven Sitzung"
            }
        }
    }

    fun startSession() {
        viewModelScope.launch {
            try {
                val now = Instant.now()
                val session = WorkSessionEntity(
                    date = LocalDate.now().toString(),
                    startTime = now.toString(),
                    isPendingSync = true,
                    isSynced = false,
                    isFinished = false
                )
                dao.insert(session)
                _activeSession.value = session
            } catch (e: Exception) {
                _error.value = "Fehler beim Starten der Sitzung"
            }
        }
    }

    fun stopSession() {
        viewModelScope.launch {
            try {
                val session = _activeSession.value ?: return@launch
                if (!session.isRunning) return@launch
                val updated = session.copy(
                    stopTime = Instant.now().toString(),
                    updatedAt = Instant.now().toString()
                )
                dao.update(updated)
                _activeSession.value = updated
            } catch (e: Exception) {
                _error.value = "Fehler beim Stoppen der Sitzung"
            }
        }
    }

    fun finishSession() {
        viewModelScope.launch {
            try {
                // If still paused, resume first to accumulate pause minutes
                if (_isPaused.value) {
                    resumeSessionInternal()
                }

                val session = _activeSession.value ?: return@launch
                val updated = session.copy(
                    stopTime = session.stopTime ?: Instant.now().toString(),
                    pauseMinutes = _accumulatedPauseMinutes.value,
                    isFinished = true,
                    isPendingSync = true,
                    updatedAt = Instant.now().toString()
                )
                dao.update(updated)
                _activeSession.value = null
                _isPaused.value = false
                _currentPauseStart.value = null
                _accumulatedPauseMinutes.value = 0
                _hasShown6h.value = false
                _hasShown9h.value = false
                _hasShown10h.value = false
                prefs?.edit()?.remove("currentPauseStart")?.apply()

                // Trigger sync after finishing session
                syncEngine?.let { engine ->
                    viewModelScope.launch {
                        try {
                            engine.syncAll()
                        } catch (e: Exception) {
                            Log.w("TimeTrackingVM", "Sync after finish failed", e)
                        }
                    }
                }
            } catch (e: Exception) {
                _error.value = "Fehler beim Abschliessen der Sitzung"
            }
        }
    }

    fun updateSession(date: String, startTime: String, stopTime: String?, pauseMinutes: Int) {
        viewModelScope.launch {
            try {
                val session = _activeSession.value ?: return@launch
                // Validation: stopTime must be after startTime
                if (stopTime != null) {
                    val start = Instant.parse(startTime)
                    val stop = Instant.parse(stopTime)
                    if (!stop.isAfter(start)) return@launch
                }
                val updated = session.copy(
                    date = date,
                    startTime = startTime,
                    stopTime = stopTime,
                    pauseMinutes = pauseMinutes,
                    isPendingSync = true,
                    updatedAt = Instant.now().toString()
                )
                dao.update(updated)
                _activeSession.value = updated
            } catch (e: Exception) {
                _error.value = "Fehler beim Aktualisieren der Sitzung"
            }
        }
    }

    fun updateFinishedSession(session: WorkSessionEntity, date: String, startTime: String, stopTime: String?, pauseMinutes: Int) {
        viewModelScope.launch {
            try {
                if (stopTime != null) {
                    val start = Instant.parse(startTime)
                    val stop = Instant.parse(stopTime)
                    if (!stop.isAfter(start)) return@launch
                }
                val updated = session.copy(
                    date = date,
                    startTime = startTime,
                    stopTime = stopTime,
                    pauseMinutes = pauseMinutes,
                    isPendingSync = true,
                    updatedAt = Instant.now().toString()
                )
                dao.update(updated)
            } catch (e: Exception) {
                _error.value = "Fehler beim Aktualisieren der Sitzung"
            }
        }
    }

    fun deleteSession(session: WorkSessionEntity) {
        viewModelScope.launch {
            try {
                dao.delete(session)
                if (session.id == _activeSession.value?.id) {
                    _activeSession.value = null
                }

                // If synced, try API DELETE immediately or queue as pending delete
                if (session.isSynced) {
                    val client = apiClient
                    if (client != null) {
                        try {
                            client.deleteWorkSession(session.id)
                        } catch (e: Exception) {
                            // API delete failed, queue as pending delete for next sync
                            database.pendingDeleteDao().insert(
                                PendingDeleteEntity(
                                    entityId = session.id,
                                    entityType = "WorkSession"
                                )
                            )
                            Log.w("TimeTrackingVM", "API delete failed, queued pending delete", e)
                        }
                    } else {
                        // No API client, queue as pending delete
                        database.pendingDeleteDao().insert(
                            PendingDeleteEntity(
                                entityId = session.id,
                                entityType = "WorkSession"
                            )
                        )
                    }
                }
            } catch (e: Exception) {
                _error.value = "Fehler beim Loeschen der Sitzung"
            }
        }
    }

    fun refresh() {
        viewModelScope.launch {
            try {
                syncEngine?.syncAll()
            } catch (e: Exception) {
                Log.w("TimeTrackingVM", "Pull-to-refresh sync failed", e)
            }
        }
    }

    // Pause methods (E08)
    fun pauseSession() {
        val session = _activeSession.value ?: return
        if (!session.isRunning || _isPaused.value) return
        _isPaused.value = true
        val now = Instant.now()
        _currentPauseStart.value = now
        prefs?.edit()?.putString("currentPauseStart", now.toString())?.apply()
    }

    fun resumeSession() {
        viewModelScope.launch {
            resumeSessionInternal()
        }
    }

    private suspend fun resumeSessionInternal() {
        if (!_isPaused.value) return
        val pauseStart = _currentPauseStart.value ?: return

        val pauseDuration = Duration.between(pauseStart, Instant.now())
        val pauseMinutes = ceil(pauseDuration.seconds / 60.0).toInt().coerceAtLeast(0)
        _accumulatedPauseMinutes.value += pauseMinutes

        val session = _activeSession.value
        if (session != null) {
            val updated = session.copy(
                pauseMinutes = _accumulatedPauseMinutes.value,
                updatedAt = Instant.now().toString()
            )
            dao.update(updated)
            _activeSession.value = updated
        }

        _isPaused.value = false
        _currentPauseStart.value = null
        prefs?.edit()?.remove("currentPauseStart")?.apply()
    }

    fun dismissArbZGBanner() {
        // The banner checks thresholds; dismissing just acknowledges the current one
    }

    fun markShown6h() { _hasShown6h.value = true }
    fun markShown9h() { _hasShown9h.value = true }
    fun markShown10h() { _hasShown10h.value = true }

    /// Create a manual session for retroactive entry (E06-S09)
    fun createManualSession(onCreated: (WorkSessionEntity) -> Unit) {
        viewModelScope.launch {
            try {
                val now = Instant.now()
                val oneHourAgo = now.minus(Duration.ofHours(1))
                val session = WorkSessionEntity(
                    date = LocalDate.now().toString(),
                    startTime = oneHourAgo.toString(),
                    stopTime = now.toString(),
                    pauseMinutes = 0,
                    isPendingSync = true,
                    isSynced = false,
                    isFinished = false
                )
                dao.insert(session)
                onCreated(session)
            } catch (e: Exception) {
                _error.value = "Fehler beim Erstellen der manuellen Sitzung"
            }
        }
    }

    fun triggerSync() {
        viewModelScope.launch {
            try {
                syncEngine?.syncAll()
            } catch (e: Exception) {
                Log.w("TimeTrackingVM", "Manual sync failed", e)
            }
        }
    }
}

File: FakturusTrack.iOS/FakturusTrack/Features/TimeTracking/TimeTrackingViewModel.swift
import SwiftData
import Foundation
import WidgetKit

@Observable @MainActor
final class TimeTrackingViewModel {
    // MARK: - State

    var activeSession: WorkSession?
    var isLoading = false
    var error: String?

    // Pause state (E08)
    var isPaused = false
    var currentPauseStart: Date?
    var accumulatedPauseMinutes: Int = 0

    // MARK: - Dependencies

    private let modelContext: ModelContext
    var syncEngine: SyncEngine?
    var networkMonitor: NetworkMonitor?

    // MARK: - UserDefaults Keys (Crash-Recovery)

    private static let pauseStartKey = "fakturus.currentPauseStart"
    private static let pauseAccumulatedKey = "fakturus.accumulatedPauseMinutes"
    private static let pauseActiveKey = "fakturus.isPaused"

    init(modelContext: ModelContext, syncEngine: SyncEngine? = nil, networkMonitor: NetworkMonitor? = nil) {
        self.modelContext = modelContext
        self.syncEngine = syncEngine
        self.networkMonitor = networkMonitor
        loadActiveSession()
        restorePauseState()
    }

    private func loadActiveSession() {
        let descriptor = FetchDescriptor<WorkSession>(
            predicate: #Predicate { !$0.isFinished }
        )
        activeSession = try? modelContext.fetch(descriptor).first
    }

    /// Restore pause state from UserDefaults after crash/restart
    private func restorePauseState() {
        guard activeSession != nil else {
            clearPauseDefaults()
            return
        }

        let defaults = UserDefaults.standard
        let wasPaused = defaults.bool(forKey: Self.pauseActiveKey)

        if wasPaused {
            isPaused = true
            accumulatedPauseMinutes = defaults.integer(forKey: Self.pauseAccumulatedKey)
            if let savedStart = defaults.object(forKey: Self.pauseStartKey) as? Date {
                currentPauseStart = savedStart
            }
        }
    }

    private func persistPauseState() {
        let defaults = UserDefaults.standard
        defaults.set(isPaused, forKey: Self.pauseActiveKey)
        defaults.set(accumulatedPauseMinutes, forKey: Self.pauseAccumulatedKey)
        if let start = currentPauseStart {
            defaults.set(start, forKey: Self.pauseStartKey)
        } else {
            defaults.removeObject(forKey: Self.pauseStartKey)
        }
    }

    private func clearPauseDefaults() {
        let defaults = UserDefaults.standard
        defaults.removeObject(forKey: Self.pauseActiveKey)
        defaults.removeObject(forKey: Self.pauseAccumulatedKey)
        defaults.removeObject(forKey: Self.pauseStartKey)
    }

    // MARK: - Actions

    func startSession() {
        let session = WorkSession(
            date: Date(),
            startTime: Date(),
            isPendingSync: true,
            isSynced: false,
            isFinished: false
        )
        modelContext.insert(session)
        try? modelContext.save()
        activeSession = session
        accumulatedPauseMinutes = 0
        isPaused = false
        currentPauseStart = nil
        clearPauseDefaults()

        SharedDefaults.writeTimerState(
            isRunning: true, startDate: session.startTime,
            isPaused: false, pauseMinutes: 0, todayTotalSeconds: 0
        )
    }

    func stopSession() {
        guard let session = activeSession, session.isRunning else { return }
        if isPaused { resumeSession() }
        session.stopTime = Date()
        session.updatedAt = Date()
        try? modelContext.save()

        SharedDefaults.writeTimerState(
            isRunning: false, startDate: session.startTime,
            isPaused: false, pauseMinutes: accumulatedPauseMinutes, todayTotalSeconds: 0
        )
    }

    func finishSession() {
        guard let session = activeSession else { return }
        if isPaused { resumeSession() }
        if session.stopTime == nil { session.stopTime = Date() }
        session.isFinished = true
        session.isPendingSync = true
        session.updatedAt = Date()
        try? modelContext.save()
        activeSession = nil
        isPaused = false
        accumulatedPauseMinutes = 0
        currentPauseStart = nil
        clearPauseDefaults()

        SharedDefaults.writeTimerState(
            isRunning: false, startDate: nil,
            isPaused: false, pauseMinutes: 0, todayTotalSeconds: 0
        )

        // Trigger sync after finishing a session
        Task { await syncEngine?.syncAll() }
    }

    func updateSession(date: Date, startTime: Date, stopTime: Date?, pauseMinutes: Int) {
        guard let session = activeSession else { return }
        if let stop = stopTime, stop <= startTime { return }

        session.date = date
        session.startTime = startTime
        session.stopTime = stopTime
        session.pauseMinutes = pauseMinutes
        session.updatedAt = Date()
        session.isPendingSync = true
        try? modelContext.save()
    }

    func deleteSession(_ session: WorkSession) {
        let wasSynced = session.isSynced
        let sessionId = session.id

        modelContext.delete(session)

        // Track pending delete for synced sessions so the server is notified
        if wasSynced {
            let pendingDelete = PendingDelete(entityId: sessionId, entityType: "WorkSession")
            modelContext.insert(pendingDelete)
        }

        try? modelContext.save()

        if sessionId == activeSession?.id {
            activeSession = nil
            isPaused = false
            accumulatedPauseMinutes = 0
            currentPauseStart = nil
            clearPauseDefaults()

            SharedDefaults.writeTimerState(
                isRunning: false, startDate: nil,
                isPaused: false, pauseMinutes: 0, todayTotalSeconds: 0
            )
        }

        // Trigger sync to process the delete on server
        if wasSynced {
            Task { await syncEngine?.syncAll() }
        }
    }

    /// Update an existing finished session (used from SessionDetailSheet)
    func updateFinishedSession(_ session: WorkSession, date: Date, startTime: Date, stopTime: Date?, pauseMinutes: Int) {
        if let stop = stopTime, stop <= startTime { return }

        session.date = date
        session.startTime = startTime
        session.stopTime = stopTime
        session.pauseMinutes = pauseMinutes
        session.updatedAt = Date()
        session.isPendingSync = true
        try? modelContext.save()

        // Trigger sync to upload the edited session
        Task { await syncEngine?.syncAll() }
    }

    /// Create a manual session (E06-S09)
    func createManualSession() -> WorkSession {
        let now = Date()
        let session = WorkSession(
            date: now,
            startTime: now.addingTimeInterval(-3600),
            stopTime: now,
            pauseMinutes: 0,
            isPendingSync: true,
            isSynced: false,
            isFinished: false
        )
        modelContext.insert(session)
        try? modelContext.save()
        return session
    }

    // MARK: - Pause (E08-S01)

    func pauseSession() {
        guard let session = activeSession, session.isRunning, !isPaused else { return }
        isPaused = true
        currentPauseStart = Date()
        persistPauseState()

        SharedDefaults.writeTimerState(
            isRunning: true, startDate: session.startTime,
            isPaused: true, pauseMinutes: accumulatedPauseMinutes, todayTotalSeconds: 0
        )
    }

    func resumeSession() {
        guard isPaused, let pauseStart = currentPauseStart else { return }
        let pauseDuration = Date().timeIntervalSince(pauseStart)
        let pauseMinutes = Int(ceil(pauseDuration / 60.0))
        accumulatedPauseMinutes += pauseMinutes

        if let session = activeSession {
            session.pauseMinutes = accumulatedPauseMinutes
            session.updatedAt = Date()
            try? modelContext.save()
        }

        isPaused = false
        currentPauseStart = nil
        persistPauseState()

        SharedDefaults.writeTimerState(
            isRunning: true, startDate: activeSession?.startTime,
            isPaused: false, pauseMinutes: accumulatedPauseMinutes, todayTotalSeconds: 0
        )
    }

    /// Computed net work minutes for ArbZG checks (accounts for current ongoing pause)
    var netWorkMinutes: Int {
        guard let session = activeSession else { return 0 }
        let totalElapsed = Date().timeIntervalSince(session.startTime)
        var totalPauseSeconds = Double(accumulatedPauseMinutes * 60)
        if let pauseStart = currentPauseStart {
            totalPauseSeconds += Date().timeIntervalSince(pauseStart)
        }
        let netSeconds = max(0, totalElapsed - totalPauseSeconds)
        return Int(netSeconds / 60)
    }
}
